// theme-provider/lib/themeEnhancer.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var actions = require("./actions");
function themeEnhancer(themeReducerName) {
    return function (next) { return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var store = next.apply(void 0, args);
        var triggerOnValueListeners = false;
        function getThemeState() {
            var state = store.getState() || {};
            if (themeReducerName)
                state = state[themeReducerName];
            return state;
        }
        var initialState = getThemeState();
        function setThemeState(state) {
            store.dispatch(actions.setState(state));
        }
        function setValue(path, value) {
            store.dispatch(actions.setValue(path, value));
        }
        function reset() {
            store.dispatch(actions.initializeState(initialState));
        }
        function clear() {
            store.dispatch(actions.initializeState({}));
        }
        return __assign({}, store, { getThemeState: getThemeState,
            setThemeState: setThemeState,
            setValue: setValue,
            reset: reset,
            clear: clear });
    }; };
}
exports.default = themeEnhancer;
